### ISSUE TITLE
Small window size on HiDPI screen

### ISSUE THREAD

**Issue by Akin909** [revery-ui/revery#226]
This was discussed a bit in #205, I'm currently working on a dell xps 15 (Manjaro Linux, gnome desktop) which has a hidpi display, the resolution is (3840 x 2160) and when opening the examples app it opens with a very small window, possibly a bug with how hidpi sizing is handled.

![examples-app](https://user-images.githubusercontent.com/22454918/51446076-de68de00-1d04-11e9-840a-fb94e53b0aad.png)

**Comment by OhadRau** [comment]
Same here on my SurfaceBook 2. (Running Windows + WSL)

**Comment by bryphe** [comment]
Thanks for the issue @Akin909 ! This helps clear things up a lot (wanted to know if it was _just_ window size, or the text size / scaling that was off too).

I think we'll need two fixes to fix this robustly across platforms:
- Example size - we'll bump up the example size (maybe 1/2 monitor resolution) just so it fills up the right space. This an easy fix.
- Font size / Scaling - this one will require a bit more work. We want the app to look / behave consistent across platforms - but in this case the app is obviously a lot smaller due to the pixel density. For some devices - they have different screen space size vs pixel size, which we detect with work like in #190 . However, in some high-DPI displays, like the Surfacebook or the Manjaro Linux size - it seems the window / framebuffer size are one-two-one - just extra small! We need to be able to detect that case and handle it (ie, scale up the font rendering, like we do for high DPI displays).

A couple of options for handling the scaling:
- There's a new API coming in GLFW 3.3 called `glfwGetWindowContentScale` which seems to be designed for this: <https://www.glfw.org/docs/3.3/window_guide.html#window_scale>. This was logged in glfw/glfw#898. However, GLFW 3.3 is unreleased now (we're on GLFW 3.2.1) - so we might need to find another way.

The patch here might give some ideas (we could potentially adapt): https://github.com/glfw/glfw/commit/16bf872117896fb88493403157fcd367303fd1f6 or there might be ways to get the scaling info.

It looks like the best way in 3.2.1 is to calculate the DPI with `glfwGetMonitorPhysicalSize`: https://www.glfw.org/docs/latest/monitor_guide.html#monitor_size

```
const double dpi = mode->width / (widthMM / 25.4);
```

And use this to decide the 'scaling'. The challenge would be just making sure the logic we choose is also compatible with the Mac logic (which has different screen / high DPI coordinates)

**Comment by bryphe** [comment]
A few things we can start looking at:

- [ ] Sample app: change size to (monitor width / 2, monitor height / 2) instead of 800x600. This still conveys the 'window-ness' but will be sized better in large monitors.
- [ ] `reason-glfw`: Implement `glfwGetMonitorPhysicalSize` stub

Once we have `glfwGetMonitorPhysicalSize`, we can use that to calculate the dpi above.

It'd be helpful to know what the values are across environments:
- The `pixelRatio` (ratio of screen size to framebuffer size)
- The `DPI` reported from `glfwGetMonitorPhysicalSize`.

It might be worth creating a simple Utility app to just output these values. I'd be interested in these particular environments:
- Mac - non-retina display
- Mac - retina display
- Linux - high DPI display
- Windows - standard DPI display
- Windows - high DPI display (Surfacebook)

It seems that the scaling is accounted for in the Mac retina display with the `pixelRatio`, however, in the Linux/Windows high-DPI display, the framebuffer + screen size seem to be 1:1, so we need to account for the scaling with another value. I believe we may need an additional value to account for DPI - like `contentScale`. The challenge is figuring out how to rationalize this with the `pixelRatio` and ensuring sizing behaves as expected across these devices.

**Comment by bryphe** [comment]
I put together a branch on top of @Akin909 's work on bryphe/reason-glfw#91 that outputs the window size, framebuffer size, and 'DPI' calculation for the native strategies. The branch is available here: https://github.com/bryphe/reason-glfw/tree/bryphe/test/dpi and just logs out these values at the beginning of the test app:

Here are the results I have so far:

| Platform | Calculated DPI | Window Size | Framebuffer Size | Notes |
| --- | --- | --- | --- | --- |
| Windows - standard DPI | `94` | `800`x`600` | `800`x`600` |  |
| Windows - Surfacebook (high DPI) | `267` | `800`x`600` | `800`x`600` | Same values for both 100%/200% scaling |
| OSX - standard DPI | `82` | `800`x`600` | `800`x`600` | |
| OSX - high DPI | `41` | `800`x`457` | `1600`x`914` | | 

@Akin909 - I'd be curious what values you get on your Linux machine? My assumption is it will behave like the Surfacebook - a high 'DPI' reported, but window size / framebuffer size match.

I'd also be interested in these values for a Macbook Pro, if someone has that available! I was running in an 'emulated' high DPI mode, so I wonder if that is impacting the results of my OSX high dpi test.

The challenge with rationalizing this is there are two cases occurring - one is the OSX case, where a framebuffer is created to match the pixel density of the monitor, but the window size is still in screen space. Today, we were using the ratio of framebuffer size / window size to calculate the device pixel ratio, but that's not sufficient for the surfacebook case - where that ratio is 1.

I suspect we might need two values:
- `pixelRatio` - the ratio between framebuffer size / window size (like we have today)
- `scaleFactor` - the ratio of the reported DPI / 96 (with a minimum value of 1).

The difference between how we handle these values is subtle - for pixelRatio, we don't want to scale the UI elements, but we do want higher-res versions. For example, if pixelRatio is 2, we'd want to render a font bitmap that is twice as large (twice as many pixels), but not take up twice as many 'screen space' pixels.

For `scaleFactor`, we'd want to increase the screen-space size. I think a simple way to accomplish this would be to divide the window size by scaleFactor 🤔 

This `scaleFactor` is not really a perfect metric, either, for example - it does not take into account the scaling setting on the surfacebook. So later, we might need a more robust way to handle this on some platforms (I think Windows has an API to get this value, which we might want to use instead of the physical DPI calculation).

**Comment by Akin909** [comment]
@bryphe I get, similar but not exactly the same value as the surface book

```
Video width: 3840
Physical width: 344
DPI (width): 283.534883721
DPI (height): 282.804123711

Framebuffer size changed: 800, 600
Window size changed: 800, 600
```

**Comment by Akin909** [comment]
Re. getting those values from the system, `gtk` has some functions for that ->
```
gdk_screen_get_monitor_scale_factor
gdk_window_get_scale_factor
gtk_widget_get_scale_factor
```

granted we can get #308 working ^^

**Comment by jchavarri** [comment]
> I'd also be interested in these values for a Macbook Pro

@bryphe here's what I see on my Macbook Pro:

| Platform | Calculated DPI | Window Size | Framebuffer Size | Notes |
| --- | --- | --- | --- | --- |
| MacOS - "Default for display" | `110.501510574` (width), `110.970873786` (height) | `800`x`600` | `1600` x `1200` | <img width="333" alt="image" src="https://user-images.githubusercontent.com/220424/52400159-a44a5b00-2abe-11e9-9267-59909dd92603.png"> |
| MacOS - "Largest text" | `78.5788519637` (width), `78.5314009662` (height) | `800`x`554` | `1600` x `1108` | <img width="436" alt="image" src="https://user-images.githubusercontent.com/220424/52400316-220e6680-2abf-11e9-927c-a823605918a2.png"> |
| MacOS - "Larger text" | `98.2235649547` (width), `98.1642512077` (height) | `800`x`600` | `1600` x `1200` | <img width="444" alt="image" src="https://user-images.githubusercontent.com/220424/52400451-74e81e00-2abf-11e9-9958-862e93bb3a94.png"> |
| MacOS - "Smaller text" | `129.309090909` (width), `129.466019417` (height) | `800`x`600` | `1600` x `1200` | <img width="420" alt="image" src="https://user-images.githubusercontent.com/220424/52400565-c1335e00-2abf-11e9-9e51-a99b371e3ac2.png"> |
| MacOS - "Smallest text" | `147.781818182` (width), `147.246376812` (height) | `800`x`600` | `1600` x `1200` | <img width="413" alt="image" src="https://user-images.githubusercontent.com/220424/52400596-df995980-2abf-11e9-9e98-dfe827466042.png"> |

**Comment by Akin909** [comment]
Fixed by #325
